### PR TITLE
sqlstats: remove unnecessary interfaces and refactor

### PR DESCRIPTION
**Pull request by mgartner**
#### insights: make Provider a struct

There's no reason to use an interface when a struct will do.

Release note: None

#### sslocal: remove unused `SQLStat.insights` field

Release note: None

#### insights: remove unused `WriterProvider` type

Release note: None

#### insights: remove `nullWriter`

The `nullWriter` has been removed. Instead, a `nil`
`*insights.ConcurrentBufferWriter` is used to disable persistence of
stats for the internal executor.

Release note: None

#### insights: remove unused `Writer` interface

Release note: None

#### insights: simplify the `Provider.Writer` method

Release note: None

#### insights: remove `sink` interface

Release note: None

#### insights: remove `insights.Reader` interface

Release note: None

#### insights: remove `LatencyInformation` interface

Epic: None

Release note: None

### CODE REVIEW

**Comment by mgartner** [comment]
Best reviewed one commit at a time :)

File: pkg/sql/sqlstats/insights/provider_test.go [hunk 1]
```diff
@@ -1,30 +0,0 @@
...
-
-import (
-	"testing"
-
-	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
-	"github.com/cockroachdb/cockroach/pkg/sql/clusterunique"
-	"github.com/stretchr/testify/require"
-)
-
-func TestProvider_WriterIgnoresInternalExecutorObservations(t *testing.T) {
```
Comment (xinhaoz) [inline pkg/sql/sqlstats/insights/provider_test.go:21]: This test doesn't seem like it was particularly useful for testing the real behaviour 🤔
I filed https://github.com/cockroachdb/cockroach/issues/128220 as a follow up for the obs team.

**Review by xinhaoz — APPROVED** [review]
Much appreciated cleanup, thank you!

File: pkg/sql/sqlstats/insights/provider_test.go (hunk 1 above)
Comment (mgartner) [reply pkg/sql/sqlstats/insights/provider_test.go:21]: Ya, it was a simple test for the `nullWriter` that didn't test much, so I removed it when I removed the `nullWriter`. I agree that an integration test would be great.

**Comment by mgartner** [comment]
TFTR!

bors r+

**Pull request merged**